thetis/interpolation.py

```diff
@@ -0,0 +1,634 @@
+    """
+    @abstractmethod
+    def __call__(self, filename, time_index):
+        """
+        Reads a data for one time step from the file
+
+        :arg str filename: a filename where to find the data (e.g. filename)
+        :arg int time_index: time index to read
+        :return: a list of floats or numpy.array_like objects
+        """
+        pass
+
+
+class NetCDFTimeSeriesReader(FileTreeReader):
+    """
+    A simple netCDF reader that returns a time slice of the given variable.
+
+    This class does not interpolate the data in any way. Useful for
+    interpolating time series.
+    """
+    def __init__(self, variable_list, time_variable_name='time'):
+        self.variable_list = variable_list
+        self.time_variable_name = time_variable_name
+        self.time_dim = None
+        self.ndims = None
+
+    def _detect_time_dim(self, ncfile):
+        assert self.time_variable_name in ncfile.dimensions
+        nc_var = ncfile[self.variable_list[0]]
+        assert self.time_variable_name in nc_var.dimensions
+        self.time_dim = nc_var.dimensions.index(self.time_variable_name)
+        self.ndims = len(nc_var.dimensions)
+
+    def _get_slice(self, time_index):
+        """
+        Returns a slice object that extracts a single time index
+        """
+        if self.ndims == 1:
+            return time_index
+        slice_list = [slice(None, None, None)]*self.ndims
+        slice_list[self.time_dim] = slice(time_index, time_index+1, None)
+        return slice
+
+    def __call__(self, filename, time_index):
+        """
+        Reads a time_index from the data base
+
+        :arg str filename: netcdf file where to find the data
+        :arg int time_index: time index to read
+        :return: a float or numpy.array_like value
+        """
+        assert os.path.isfile(filename), 'File not found: {:}'.format(filename)
+        with netCDF4.Dataset(filename) as ncfile:
+            if self.time_dim is None:
+                self._detect_time_dim(ncfile)
+            output = []
+            for var in self.variable_list:
+                values = ncfile[var][self._get_slice(time_index)]
+                output.append(values)
+            return output
+
+
+class SpatialInterpolator2d(object):
+    """
+    Abstract spatial interpolator class that can interpolate onto a Function
+    """
+    def __init__(self, function_space, to_latlon):
+        """
+        :arg function_space: target Firedrake FunctionSpace
+        :arg to_latlon: Python function that converts local mesh coordinates to
+            latitude and longitude: 'lat, lon = to_latlon(x, y)'
+        """
+        self.function_space = function_space
+
+        # construct local coordinates
+        xy = SpatialCoordinate(self.function_space.mesh())
+        fsx = Function(self.function_space).interpolate(xy[0]).dat.data_with_halos
+        fsy = Function(self.function_space).interpolate(xy[1]).dat.data_with_halos
```

Comment:
You can destructure `x, y = SpatialCoordinate(...)`.  Should probably `assert function_space.ufl_element().value_shape() == ()`